awx/main/models/events.py

```diff
@@ -0,0 +1,685 @@
+
+    def save(self, *args, **kwargs):
+        # If update_fields has been specified, add our field names to it,
+        # if it hasn't been specified, then we're just doing a normal save.
+        update_fields = kwargs.get('update_fields', [])
+        # Update model fields and related objects unless we're only updating
+        # failed/changed flags triggered from a child event.
+        from_parent_update = kwargs.pop('from_parent_update', False)
+        if not from_parent_update:
+            # Update model fields from event data.
+            updated_fields = self._update_from_event_data()
+            for field in updated_fields:
+                if field not in update_fields:
+                    update_fields.append(field)
+            # Update host related field from host_name.
+            if not self.host_id and self.host_name:
+                host_qs = self.job.inventory.hosts.filter(name=self.host_name)
+                host_id = host_qs.only('id').values_list('id', flat=True).first()
+                if host_id != self.host_id:
+                    self.host_id = host_id
+                    if 'host_id' not in update_fields:
+                        update_fields.append('host_id')
+        super(JobEvent, self).save(*args, **kwargs)
+        # Update related objects after this event is saved.
+        if not from_parent_update:
+            if getattr(settings, 'CAPTURE_JOB_EVENT_HOSTS', False):
+                self._update_hosts()
+            if self.event == 'playbook_on_stats':
+                self._update_parents_failed_and_changed()
+
+                hostnames = self._hostnames()
+                self._update_host_summary_from_stats(hostnames)
+                self.job.inventory.update_computed_fields()
+
+                emit_channel_notification('jobs-summary', dict(group_name='jobs', unified_job_id=self.job.id))
+
+    @classmethod
+    def create_from_data(self, **kwargs):
+        # Must have a job_id specified.
+        if not kwargs.get('job_id', None):
+            return
+
+        # Convert the datetime for the job event's creation appropriately,
+        # and include a time zone for it.
+        #
+        # In the event of any issue, throw it out, and Django will just save
+        # the current time.
+        try:
+            if not isinstance(kwargs['created'], datetime.datetime):
+                kwargs['created'] = parse_datetime(kwargs['created'])
+            if not kwargs['created'].tzinfo:
+                kwargs['created'] = kwargs['created'].replace(tzinfo=utc)
+        except (KeyError, ValueError):
+            kwargs.pop('created', None)
+
+        # Sanity check: Don't honor keys that we don't recognize.
+        valid_keys = {'job_id', 'event', 'event_data', 'playbook', 'play',
+                      'role', 'task', 'created', 'counter', 'uuid', 'stdout',
+                      'parent_uuid', 'start_line', 'end_line', 'verbosity'}
```

Comment:
I wish we could make this something other than a manually curated set, but I see why it's hard.

If you take the fields present on the JobEvent model which are not in this list (effectively the excluded fields) you get

```
{'changed', 'failed', 'host', 'host_name', u'id', 'job', 'modified', 'parent'}
```

So these are kind of the managed fields for the most part. AWX code itself will set things like the modified time and the parent and host_name. So there aren't any real good options to cut this down.

Reply:
Yea, I think this is fine; it sort of is what it is.